File: baekjoonn/2565_전깃줄/B_2565_2.js
const binarySearch = function(start, end, num) {
    if (start === end) {
        return end
    }

    const mid = Math.floor((start + end) / 2)

    if (num < dp[mid]) {
        return binarySearch(start, mid, num)
    }

    if (num > dp[mid]) {
        return binarySearch(mid + 1, end, num)
    }
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const N = parseInt(input[0])
let electricCords = []

for (let i = 1; i <= N; i++) {
    const cord = input[i].split(' ').map(v => parseInt(v))
    electricCords.push(cord)
}

const sortedElectricCords = electricCords.sort((a, b) => {
    // if (a[0] === b[0]) {
    //     return a[1] - b[1]
    // }
    // 같은 위치에 두 개 이상의 전깃줄이 연결될 수 없다고 했으므로
    // 위의 조건은 없어도 된다
    return a[0] - b[0]
})

let dp = [sortedElectricCords[0][1]]
for (let i = 1; i < N; i++) {
    const [a, b] = sortedElectricCords[i]
    if (b > dp[dp.length - 1]) {
        dp.push(b)
    } else {
        const idx = binarySearch(0, dp.length - 1, b)
        dp[idx] = b
    }
}

console.log(N - dp.length)

File: baekjoonn/2565_전깃줄/B_2565_2.py
def binary_search(start, end, target):
    while start < end:
        mid = (start + end) // 2

        if lis[mid] >= target:
            end = mid
        else:
            start = mid + 1

    return start


N = int(input())
wires = [list(map(int, input().split())) for _ in range(N)]
wires.sort(key=lambda x: x[0])

lis = [wires[0][1]]

for i in range(1, N):
    if lis[-1] < wires[i][1]:
        lis.append(wires[i][1])
    else:
        # end 값에 주의
        idx = binary_search(0, len(lis) - 1, wires[i][1])
        lis[idx] = wires[i][1]

print(N - len(lis))

# https://www.acmicpc.net/board/view/8218
# 10
# 1 6
# 2 8
# 3 2
# 4 9
# 5 5
# 6 10
# 7 4
# 8 1
# 9 7
# 10 3
# => 6
